onnxscript/rewriter/ort_fusions/fuse_mha_bias.py

```diff
@@ -0,0 +1,203 @@
+            num_heads=num_heads,
+            _domain="com.microsoft",
+        )
+
+    def check(
+        self,
+        op,
+        query_matmul,
+        key_matmul,
+        value_matmul,
+        **_,
+    ) -> pattern.MatchResult:  # type: ignore[name-defined]
+        check_result = pattern.MatchResult()
+
+        self.bindings: dict[str, Dim] = {}
+
+        def no_match(val: ir.Value, dims: Sequence[str]) -> bool:
+            return not _fusion_utils._check_shape(self.bindings, val, dims)
+
+        if no_match(query_matmul, ["B", "S", "D"]):
+            return check_result.fail(
+                f"Shape mismatch: {query_matmul} does not match expected dimensions ['B', 'S', 'D']",
+                query_matmul,
+            )
+        if no_match(key_matmul, ["B", "Skv", "Dk"]):
+            return check_result.fail(
+                f"Shape mismatch: {key_matmul} does not match expected dimensions ['B', 'Skv', 'Dk']",
+                key_matmul,
+            )
+        if no_match(value_matmul, ["B", "Skv", "Dv"]):
+            return check_result.fail(
+                f"Shape mismatch: {value_matmul} does not match expected dimensions ['B', 'Skv', 'Dv']",
+                value_matmul,
+            )
+
+        self.Dh_q = self.bindings.get("D")
+        self.Dh_k = self.bindings.get("Dk")
+        self.Dh_v = self.bindings.get("Dv")
+
+        if (
+            not isinstance(self.Dh_q, int)
+            or not isinstance(self.Dh_k, int)
+            or not isinstance(self.Dh_v, int)
+        ):
+            return check_result.fail(
+                "Could not determine the hidden sizes of query, key, and value.",
+            )
+
+        return check_result
+
+    def rewrite(
+        self,
+        op,
+        query_matmul,
+        key_matmul,
+        value_matmul,
+        q_bias,
+        k_bias,
+        v_bias,
+        mask,
+        past_key,
+        past_value,
+        num_heads,
+        **_,
+    ):
+        if self._q_no_bias:
+            q_bias = op.Constant(
+                value=ir.tensor(numpy.zeros((self.Dh_q,), dtype=numpy.float32))
+            )
+        if self._k_no_bias:
+            k_bias = op.Constant(
+                value=ir.tensor(numpy.zeros((self.Dh_k,), dtype=numpy.float32))
+            )
+        if self._v_no_bias:
+            v_bias = op.Constant(
+                value=ir.tensor(numpy.zeros((self.Dh_v,), dtype=numpy.float32))
+            )
+        bias = op.Concat(q_bias, k_bias, v_bias, axis=0)
+        return op.MultiHeadAttention(
+            query_matmul,
+            key_matmul,
+            value_matmul,
+            bias,
+            None,
+            mask,
+            past_key,
+            past_value,
+            num_heads=num_heads,
+            _domain="com.microsoft",
+        )
+
+
+parameter_combinations = [
+    {
+        "q_no_bias": q_no_bias,
+        "k_no_bias": k_no_bias,
+        "v_no_bias": v_no_bias,
+    }
+    for q_no_bias in [False, True]
```

Comment:
The (True, True, True) case seems unnecessary? Won't it unnecessary introduce a zero-valued bias and add it? Seems inefficient.

Reply:
Set bias=None for this case.

Reply:
But does the rule do anything in the (True, True, True) case? Based on my reading, it does nothing. So, why not exclude this combination from the rules in the first place? You just need to add a `if not (q_no_bias and k_no_bias and v_no_bias)` down below?